cirq/protocols/decompose.py

```diff
@@ -0,0 +1,234 @@
+        on_stuck_raise: If there is an operation that can't be decomposed and
+            also can't be kept, `on_stuck_raise` is used to determine what error
+            to raise. `on_stuck_raise` can either directly be an `Exception`, or
+            a method that takes the problematic operation and returns an
+            `Exception`. If `on_stuck_raise` is set to `None`, undecomposable
+            operations are simply silently kept. `on_stuck_raise` defaults to a
+            `ValueError` describing the unwanted undecomposable operation.
+
+    Returns:
+        A list of operations that the given value was decomposed into. If
+        `on_stuck_raise` isn't set to None, all operations in the list will
+        satisfy the predicate specified by `keep`.
+
+    Raises:
+        TypeError:
+            `val` isn't a `cirq.Operation` and can't be decomposed even once.
+            (So it's not possible to return a list of operations.)
+
+        ValueError:
+            Default type of error raised if there's an undecomposable operation
+            that doesn't satisfy the given `keep` predicate.
+
+        TError:
+            Custom type of error raised if there's an undecomposable operation
+            that doesn't satisfy the given `keep` predicate.
+    """
+    from cirq import ops  # HACK: Avoids circular dependencies.
+
+    if on_stuck_raise is not None and keep is None:
+        raise ValueError(
+            "Must specify 'keep' if specifying 'on_stuck_raise', because it's "
+            "not possible to get stuck if you don't have a criteria on what's "
+            "acceptable to keep.")
+
+    if intercepting_decomposer is None:
+        actual_decomposer = decomposer
+    else:
+        actual_decomposer = lambda op: _intercept_decompose(
+            op, [intercepting_decomposer, decomposer])
+
+    output = []
+    queue = []  # type: List['cirq.Operation']
+    if not isinstance(val, ops.Operation):
+        queue.extend(decompose_once(val))
+    else:
+        queue.append(val)
+
+    while queue:
+        item = queue.pop(0)
+
+        if keep is not None and keep(item):
+            output.append(item)
+            continue
+
+        decomposed = actual_decomposer(item)
+        if decomposed is not NotImplemented:
+            queue.extend(ops.flatten_op_tree(decomposed))
+            continue
+
+        if isinstance(item, collections.Iterable):
+            queue.extend(ops.flatten_op_tree(item))
+            continue
+
+        if on_stuck_raise is not None:
+            if isinstance(on_stuck_raise, Exception):
+                raise on_stuck_raise
+            else:
+                raise on_stuck_raise(item)
+
+        output.append(item)
+
+    return output
+
+
+def decompose_once(val: Any,
+                   default: TDefault=RaiseTypeErrorIfNotProvided,
+                   **kwargs
+                   ) -> Union[TDefault, List['cirq.Operation']]:
+    """Decomposes a value into operations, if possible.
+
```

Comment:
doc to that this is not recursive for completeness

Reply:
Done.